Update documentation (#658)
ethereum/solc-js: README.md

For examples how to use them, please refer to the README of the above mentioned solc-js releases.

**Note**: These low-level functions remain available for compatibility reasons.
However, they were superseded by the `compile()` function and are no longer required.

> Starting in version `0.5.0+commit.1d4f565a`, the functions `compileSingle`, `compileMulti`, and `compileCallback` are always `null` when using newer binary versions.

Reviewer: ```suggestion
Starting from version `0.5.0+commit.1d4f565a`, the functions `compileSingle`, `compileMulti`, and `compileCallback` are always `null` when using newer binary versions.
```
Also, I find the last part, `when using newer binary versions`, a little bit confusing because it doesn't specify which versions are `newer` and that those are `solc` binaries, not `solc-js` (right?).

Author: Right. Fixed.

> It is recommended to use the latest release of solc-js, but it should also handle all the older binaries down to `0.1.x`.

Reviewer: ```suggestion
It is recommended to use the latest release of solc-js, but it should also handle all the older solc binaries down to `0.1.x`.
```
This feels confusing to me too, who should handle older binaries? `solc` or `solc-js` binaries?

Author: The `solc-js` should handle old `solc` binaries.

For examples how to use them, please refer to the README of the above mentioned solc-js releases.

> **Note**: These low-level functions still present for compatibility reasons to support all compiler binaries, but they were superseded by the `compile()` function and are no longer required in newer versions. The functions `compileSingle`, `compileMulti` and `compileCallback` are always `null` in the newer versions.

Reviewer 2: I'm no native English speaker, but using `present` here sounds somehow weird to me. Why not "These low-level functions remain available for compatibility reasons." IMHO a simpler sentence.

Author: I will change it :)

For examples how to use them, please refer to the README of the above mentioned solc-js releases.

**Note**: These low-level functions remain available for compatibility reasons.

> However, they were superseded by the `compile()` function and are no longer required in newer versions. The functions `compileSingle`, `compileMulti` and `compileCallback` are always `null` in the newer versions.

Reviewer 3: Please clarify that they're `null` when you use a newer binary (might also be good to check which version specifically; was that 0.5.0?) rather than newer solc-js. This distinction might not be clear to users.

For some context: it is recommended to always use latest release of solc-js and it's supposed to be able to handle handle all the older binaries, down to 0.1.x. Might be good to mention that in the README too if it's not mentioned already.

Author: Yes, `0.5.0+commit.1d4f565a`.

    // An error was encountered, display and quit
  } else {
    // NOTE: Use `solcSnapshot` here with the same interface `solc` has
    // For example:
    const output = solcSnapshot.compile(/* ... */)
  }
});
```

> The version should be in the long format string.

Reviewer 2: "should" or "must"?

Author: `must`. Otherwise, it will not find the binary.

    // An error was encountered, display and quit
  } else {
    // NOTE: Use `solcSnapshot` here with the same interface `solc` has
    // For example:
    const output = solcSnapshot.compile(/* ... */)
  }
});
```

> The version **must** be in the long format string.

Reviewer 3: ```suggestion
The version **must** be in the long format.
```

> Thus, if you would like to use version `v0.8.17` you need to provide it, including the committed version of the release.

Reviewer 3: ```suggestion
Thus, if you would like to use version `v0.8.17` you need to include the commit hash of the release.
```

> You can extract the long version string for each version from the [public available release list](https://binaries.soliditylang.org/bin/list.json).

Reviewer 3: I wanted to say that we should rather point people towards the new `emscripten-wasm32/` and `emscripten-asmjs/` dirs but I realized that it would be inconsistent because we're still using and mentioning `bin/` everywhere. So I guess it's out of scope of this PR. FYI I created an issue for it: #671.

    // An error was encountered, display and quit
  } else {
    // NOTE: Use `solcSnapshot` here with the same interface `solc` has
    // For example:
    const output = solcSnapshot.compile(/* ... */)
  }
});
```

The version **must** be in the long format.
Thus, if you would like to use version `v0.8.17` you need to include the commit hash of the release.

> You can extract the long version string for each version from the [public available release list](https://binaries.soliditylang.org/bin/list.json).

Reviewer 3: ```suggestion
You can extract the long version string for each version from the [publicly available release list](https://binaries.soliditylang.org/bin/list.json).
```

```html
<script
  type="text/javascript"
  src="https://binaries.soliditylang.org/bin/{{ SOLC_VERSION }}.js"
  integrity="sha256-{{ BASE64_ENCODED_HASH_OF_SOLC_VERSION  }}"
  crossorigin="anonymous"
></script>
```

This will load `solc` into the global variable `window.Module`.

> Alternatively, use `https://binaries.soliditylang.org/bin/soljson-latest.js` to get the latest version.

Reviewer: I thought about something like [latest version script](https://binaries.soliditylang.org/bin/soljson-latest.js), for example. Do you think this is appropriate?

> Alternatively, use `https://binaries.soliditylang.org/bin/soljson-latest.js` to get the latest version.

Author: Oh right, my idea was to be more explicit about the location of the latest version since in the example above, it is shown as: `https://binaries.soliditylang.org/bin/{{ SOLC_VERSION }}.js`. And the link refers to an alternative to the example. But yes, we can omit the link if it makes it easier to understand.

Or maybe just: "Alternatively, replace `{{ SOLC_VERSION }}.js` by `soljson-latest.js` in the link above...`

> Alternatively, use `https://binaries.soliditylang.org/bin/soljson-latest.js` to get the latest version.

Reviewer: Understood. IMO, the latter suggestion is better, but just because I think that in general hiding the URLs behind an hyperlink is more elegant.

> Alternatively, use `https://binaries.soliditylang.org/bin/soljson-latest.js` to get the latest version.

Author: But which latter suggestion? Yours or mine?
I think it should be clear that what we are referring to is the URL to be used in the script tag in the example. What about this one?

"Alternatively, use [soljson-latest.js](https://binaries.soliditylang.org/bin/soljson-latest.js) as `{{ SOLC_VERSION }}.js` in the code snippet above to load the latest version."

> Alternatively, use `https://binaries.soliditylang.org/bin/soljson-latest.js` to get the latest version.

Reviewer: I would suggest
```
Alternatively, use [soljson-latest.js](https://binaries.soliditylang.org/bin/soljson-latest.js) instead of {{ SOLC_VERSION }}.js in the code snippet above to load the latest version.
```

> Alternatively, use `soljson-latest.js` as `{{ SOLC_VERSION }}.js` in the code snippet above to load the latest version.

Reviewer: ```suggestion
Alternatively, use `soljson-latest.js` instead of `{{ SOLC_VERSION }}.js` in the code snippet above to load the latest version.
```

It is recommended that you check the integrity of the resource being fetched before using it in your application.
For that, you can use the [Subresource Integrity (SRI)](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) feature.
Adding SRI configuration to your HTML script tag ensures that the resource will only be loaded by the browser if the cryptographic hashes match.

You can generate the SRI hash yourself based on the base64-encoded version of the sha256 hash of the release.
For example, after downloading version `v0.8.16+commit.07a7930e`, run:
```bash

> shasum -b -a 256 soljson-v0.8.16+commit.07a7930e.js | awk '{ print $1 }' | xxd -r -p | base64

Reviewer 3: I think `xxd` may be a bit obscure (on my system it comes with gvim :)). Personally, I do use it, but for our users, who are likely JS developers perhaps it would be better to suggest a JS one-liner via `node -e` instead?

Author: > on my system it comes with gvim
Really? Weird! haha

Ok. I think that the use of `xxd` is fine for the sake of an example. But I updated it to use node as you suggested.

> shasum -b -a 256 soljson-v0.8.16+commit.07a7930e.js | awk '{ print $1 }' | xxd -r -p | base64

Reviewer 3: Any reason not to use `sha256sum` instead? Makes it a bit more compact and easier to type.
```suggestion
sha256sum --binary soljson-v0.8.16+commit.07a7930e.js | awk '{ print $1 }' | xxd -revert -plain | base64
```

Author: No, no particular reason. Changed.

> ```bash
> shasum -b -a 256 soljson-v0.8.16+commit.07a7930e.js | awk '{ print $1 }' | xxd -r -p | base64
> J7KCDvk4BaZcdreUWklDJYLTBv0XoomFcJpR5kA2d8I=
> ```

Reviewer 3: This looks as if the hash was a part of the command to run.
````suggestion
```bash
shasum -b -a 256 soljson-v0.8.16+commit.07a7930e.js | awk '{ print $1 }' | xxd -r -p | base64
```
```
J7KCDvk4BaZcdreUWklDJYLTBv0XoomFcJpR5kA2d8I=
```
````

<script
  type="text/javascript"
  src="https://binaries.soliditylang.org/bin/{{ SOLC VERSION }}.js"

>   integrity="sha256-{{ BASE64-ENCODED HASH OF SOLC VERSION  }}"

Reviewer 2: where is that `BASE64-ENCODED HASH OF SOLC VERSION` coming from?

Author: Yes. Maybe I can put in between underscore if that makes it easier to understand that it is just a placeholder.

Reviewer 3: Honestly, for me the underscores make it look like an existing variable rather than a placeholder so I wanted to suggest removing them :P

  crossorigin="anonymous"
></script>
```

> This will load `solc` into the global variable `window.Module`.

Reviewer 2: Into the global variable or is that a namespace / object? Maybe saying assigning it as a member of `window.Module` would be more precise?

Author: I think `global variable` is correct. I mean, the `Module` is a property of window and thus a global variable since window is a global object of a web page that refers to itself. I may be wrong, but solc is not assigned as a member of the `window.Module`, it is loaded as the Module object in the window. Maybe `global object` would be correct, too. Like:
```
This will load solc as the global object window.Module.
```

But honestly, I'm not sure about the terminology. I usually refer to any window property as global variables, but I may be calling them by the wrong name, haha

Alternatively, use `https://binaries.soliditylang.org/bin/soljson-latest.js` to get the latest version.

It is recommended that you check the integrity of the resource being fetched before using it in your application.
For that, you can use the [Subresource Integrity (SRI)](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) feature.
Adding SRI configuration to your HTML script tag ensures that the resource will only be loaded in the browser if the cryptographic hashes matches.

You can generate the SRI hash yourself based on the base64-encoded version of the sha256 hash of the release.
For example, after downloading version `v0.8.16+commit.07a7930e`, run:
```bash
shasum -b -a 256 soljson-v0.8.16+commit.07a7930e.js | awk '{ print $1 }' | xxd -r -p | base64
J7KCDvk4BaZcdreUWklDJYLTBv0XoomFcJpR5kA2d8I=
```

And update your `index.html` to:
```html
<script
  type="text/javascript"
  src="https://binaries.soliditylang.org/bin/soljson-v0.8.16+commit.07a7930e.js"
  integrity="sha256-J7KCDvk4BaZcdreUWklDJYLTBv0XoomFcJpR5kA2d8I="
  crossorigin="anonymous"
></script>
```

> Then use this inside Javascript as:

Reviewer 2: ```suggestion
Then use this inside JavaScript as:
```
I think that's CamelCase, isn't it?
